autopilot/contractor.go

```diff
@@ -516,10 +516,12 @@ func (c *contractor) runContractChecks(ctx context.Context, w Worker, contracts
 			continue
 		}
 
-		// starting here we need a revision for checking the contract. So if
-		// there is no revision, the contract isn't considered good.
+		// check if we have a revision
 		if contract.Revision == nil {
-			toStopUsing[fcid] = errContractNoRevision.Error()
+			// NOTE: if we can not fetch the revision, we give the host the
+			// benefit of the doubt and keep it in the contractset, removing it
+			// might cause a lot of unnecessary migrations
+			toKeep = append(toKeep, fcid)
```

Comment:
I think the way this is now potentially causes very bad sets.

1. both places where we now add a contract to `toKeep` if they don't have a revision happen before the `isOffline` check within `isUsableHost`. So we might add offline contracts to the set.
2. hosts that are gouging might also be added to the set for the same reason.

I think we should be a bit more restrictive here doing some of the following:
- You only get the benefit of the doubt if you were in the contract set before
- We limit the number of contracts without revision in the set
- You only get the benefit of the doubt a limited number of times

Reply:
Another edge case is the fact that a contract without revision has a size of 0. So it might get truncated even if it is actually a pretty big contract.